enhancements/sig-architecture/229-manifestworkreplicaset-rollback/README.md

```diff
@@ -0,0 +1,425 @@
+
+## Summary
+
+This enhancement proposes adding rollback capabilities to ManifestWorkReplicaSet (MWRS) to enable safe recovery from failed multi-cluster deployments. The solution leverages Kubernetes ControllerRevision resources to maintain a historical record of MWRS template changes, similar to how StatefulSet and DaemonSet controllers track revision history. 
+
+The enhancement introduces two key rollback mechanisms: automatic rollback on failure (via `abortOnFailure` flag) and manual rollback (via spec updates). When a progressive rollout fails, the controller can automatically revert to the previous stable revision across all affected clusters, minimizing downtime and impact. For manual rollback scenarios, users can update the MWRS spec to reference a previous revision, and the controller will restore that revision.
+
+New API fields include `revisionHistoryLimit` to control history retention, `abortOnFailure` to enable automatic rollback, and status fields (`currentRevision`, `updateRevision`, `abort`, `abortedTime`, `collisionCount`) to track rollout state. The enhancement also introduces a new `Progressing` condition type and `RolloutDegraded` reason to provide better visibility into rollout and rollback operations.
+
+## Motivation
+
+The ManifestWorkReplicaSet (MWRS) simplifies multi-cluster workload distribution by supporting several rollout strategies that enable gradual deployment of workloads across clusters. This allows operators to safely deploy applications to multiple managed clusters with controlled rollout patterns.
+
+However, the current implementation has limitation: when a progressive rollout fails, the Work controller simply stops the rollout, leaving clusters in an inconsistent state. Some clusters may have the new version deployed while others remain on the old version, and the failed clusters are stuck in a failed state. This creates several problems:
+
+1. **Inconsistent Cluster State**: Partial rollouts leave the system in an undefined state where different clusters run different versions
+2. **Extended Downtime**: Critical applications may experience extended downtime while waiting for manual intervention
+
+For critical applications, especially those requiring high availability, the ability to automatically roll back to a previous stable revision on failure is essential to minimize impact and maintain service continuity. This enhancement addresses these limitations by providing both automatic and manual rollback capabilities with proper revision history management.
+
+### Goals
+
+- Define manifestwork template revision management.
+- Enable rollback feature with the revision.
+- Enable automatic rollout on the failure of rollout.
+
+### Non-Goals
+
+- Manual rollback via kubectl plugin commands (rollback will be performed by updating `.spec.manifestWorkTemplate`)
+- Rollback of CRD versions (automatic rollback should be used with caution when CRDs are involved)
```

Comment:
Does this mean CRDs deployed by MWRS will not be included in the rollback?

Reply:
It will support any resources, however, we will need to add the caution in the document in the future. I would delete this part.